cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -1604,19 +1623,145 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
+                        } else {
+                            log.debug("{}: Determining whether kafka broker {} can be update dynamically", reconciliation, finalPodId);
+                            Promise<Void> p = Promise.promise();
+                            brokerConfig(ac, finalPodId)
+                                    .compose(brokerConfig -> {
+                                        log.trace("{}: Broker description {}", reconciliation, brokerConfig);
+                                        KafkaBrokerConfigurationDiff configurationDiff = new KafkaBrokerConfigurationDiff(brokerConfig, this.kafkaConfig, kafkaCluster.getKafkaVersion(), finalPodId);
+                                        if (!configurationDiff.isEmpty()) {
+                                            if (!configurationDiff.canBeUpdatedDynamically()) {
+                                                kafkaPodsUpdatedDynamically.put(finalPodId, "dynamic update not possible");
+                                                return Future.succeededFuture();
+                                            } else {
+                                                return updateBrokerConfigDynamically(finalPodId, ac, configurationDiff);
+                                            }
+                                        } else {
+                                            log.debug("{}: Kafka broker {} configuration is unchanged", reconciliation, finalPodId);
+                                            kafkaPodsUpdatedDynamically.put(finalPodId, "");
+                                        }
+                                        return Future.succeededFuture();
+                                    }).onComplete(ign -> {
+                                        if (ac != null) {
+                                            try {
+                                                ac.close(Duration.ofMinutes(2));
+                                            } catch (Throwable e) {
+                                                log.warn("{}: Ignoring exception when closing admin client", reconciliation, e);
+                                            }
+                                        }
+                                        p.handle(ign);
+                                    });
+                            return p.future();
+                        }
+                    }).recover(ignore2 -> {
+                        log.debug("{}: recovering from the failed dynamic updating {}", reconciliation, ignore2);
+                        kafkaPodsUpdatedDynamically.put(finalPodId, "dynamic update failed");
+                        return Future.succeededFuture();
+                    })).recover(fail -> {
+                        log.warn("{}: kafka pod {} not ready", reconciliation, finalPodId);
+                        kafkaPodsUpdatedDynamically.put(finalPodId, "");
+                        return Future.succeededFuture();
+                    }));
```

Comment:
It's very difficult to see that this `recover()` recovering from the readiness check. It looks like it's recovering from the `recover()` above. Can we do something with indent to make it clearer?